### DISCUSSION TITLE
How do i enable nullAnalysis?

### DISCUSSION THREAD

**Discussion by alven-shin** [mfussenegger/nvim-jdtls#522] [Q&A]
I would like to enable null analysis so I get warnings when I forget null checks on annotated variables. I tried the following config but it does not seem to work.

```lua
-- ftplugin/java.lua
local config = {
  cmd = { "jdtls" },
  root_dir = vim.fs.dirname(vim.fs.find({ "gradlew", ".git", "mvnw" }, { upward = true })[1]),
  settings = {
    java = {
      compile = {
        nullAnalysis = {
          mode = "automatic",
        },
      },
    },
  },
}
require("jdtls").start_or_attach(config)
```
I am testing the feature with the following code, and in VS Code with the Java extension, a warning is produced for the missing null check.

```java
import javax.annotation.Nullable;

public class Test {
    private @Nullable String name;

    public void print() {
        System.out.println(this.name.length()); // Warning only appears on VSCode: Potential null pointer access: The field name is specified as @Nullable Java(33555356)
    }
}
```

**Comment by agriffis** [comment]
Did you ever figure this out? I had a similar question at https://github.com/eclipse-jdtls/eclipse.jdt.ls/discussions/2952 but I could never understand how to apply the suggestion.

**Reply by alven-shin** [reply]
I did not. The last time i tried was about a year ago, so things may have changed since then. There is also the [nvim-java](https://github.com/nvim-java/nvim-java) plugin now which may work better.

**Reply by agriffis** [reply]
Thanks for the reply! I have made some progress at https://github.com/eclipse-jdtls/eclipse.jdt.ls/issues/2655#issuecomment-2230843714

**Comment by notFoxils** [comment]
For anyone looking in the modern day.
```lua
settings = {
    java = {
        compile = {
            nullAnalysis = {
                nonnull = {
                    "javax.annotation.Nonnull",
                    "org.eclipse.jdt.annotation.NonNull",
                    "org.springframework.lang.NonNull",
                    "org.jspecify.annotations.NonNull"
                },
                nullable = {
                    "javax.annotation.Nullable",
                    "org.eclipse.jdt.annotation.Nullable",
                    "org.springframework.lang.Nullable",
                    "org.jspecify.annotations.Nullable"
                },
                nonnullbydefault = {
                    "javax.annotation.ParametersAreNonnullByDefault",
                    "org.eclipse.jdt.annotation.NonNullByDefault",
                    "org.springframework.lang.NonNullApi",
                    "org.jspecify.annotations.NullMarked"
                },
                mode = "automatic",
            }
        }
    }
}
```
(likely to be the same issues initially encountered above, I'm not sure how different jdt.ls is from then to now, nor if there were any compounded issues)
The reason it hadn't worked for the unfortunate soul above, and more specifically for me was because the `nonnull`, `nullable`, `nonnullbydefault` arrays need to be set and found in the project's classpath.

If any of them are not found, or if any of the lists are unset, the default eclipse annotations are used, and the null analysis is disabled.

- [classpath-searching](https://github.com/eclipse-jdtls/eclipse.jdt.ls/blob/f77624d6f0502c878527c474c7b5b09efc804e81/org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/preferences/Preferences.java#L3214-L3237)
- [any-not-found](https://github.com/eclipse-jdtls/eclipse.jdt.ls/blob/f77624d6f0502c878527c474c7b5b09efc804e81/org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/preferences/Preferences.java#L3345)
- [null-analysis-disabled](https://github.com/eclipse-jdtls/eclipse.jdt.ls/blob/f77624d6f0502c878527c474c7b5b09efc804e81/org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/preferences/Preferences.java#L3346)
- [defaults-getting-set](https://github.com/eclipse-jdtls/eclipse.jdt.ls/blob/f77624d6f0502c878527c474c7b5b09efc804e81/org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/preferences/Preferences.java#L3349-L3351)
- default-annotations
  - [nullable](https://help.eclipse.org/latest/topic/org.eclipse.jdt.doc.isv/reference/api/org/eclipse/jdt/core/JavaCore.html#COMPILER_NULLABLE_ANNOTATION_NAME)
  - [nonnull](https://help.eclipse.org/latest/topic/org.eclipse.jdt.doc.isv/reference/api/org/eclipse/jdt/core/JavaCore.html#COMPILER_NONNULL_ANNOTATION_NAME)
  - [nonnull-by-default](https://help.eclipse.org/latest/topic/org.eclipse.jdt.doc.isv/reference/api/org/eclipse/jdt/core/JavaCore.html#COMPILER_NONNULL_BY_DEFAULT_ANNOTATION_NAME)

***sigh***
